### PR TITLE
Lids

### PR DESCRIPTION
**Pull request by bmwshop**
# What does this PR do ?

Enable optional output of language IDs during AASR transcription / inference

**Collection**: [Note which collection this PR will affect]
ASR 

# Changelog 
- Aggregate tokenizer modified so that it can output token and sample LIDs
- rnnt_wer.py modified to handle the compute_langs flag and produce output
- rnnt_wer_bpe.py modified to handle the compute_langs flag and produce output

# Usage

See the sample code below. This works similar to timestamp output.
```
# Run this:
# python langs.py

# Sample output
Transcription:   o que es demasiado ruidosa o que tiene palabras que no deberían de ponerlas en canciones porque es como

Langs:  es

Langs chars [{'char': ' o', 'lang': 'es'}, {'char': ' que', 'lang': 'es'}, {'char': ' es', 'lang': 'es'}, {'char': ' de', 'lang': 'es'}, {'char': 'ma', 'lang': 'es'}, {'char': 's', 'lang': 'es'}, {'char': 'i', 'lang': 'es'}, {'char': 'ado', 'lang': 'es'}, {'char': ' ru', 'lang': 'es'}, {'char': 'ido', 'lang': 'es'}, {'char': 's', 'lang': 'es'}, {'char': 'a', 'lang': 'es'}, {'char': ' o', 'lang': 'es'}, {'char': ' que', 'lang': 'es'}, {'char': ' tiene', 'lang': 'es'}, {'char': ' palabra', 'lang': 'es'}, {'char': 's', 'lang': 'es'}, {'char': ' que', 'lang': 'es'}, {'char': ' no', 'lang': 'es'}, {'char': ' debe', 'lang': 'es'}, {'char': 'ría', 'lang': 'es'}, {'char': 'n', 'lang': 'es'}, {'char': ' de', 'lang': 'es'}, {'char': ' poner', 'lang': 'es'}, {'char': 'las', 'lang': 'es'}, {'char': ' en', 'lang': 'es'}, {'char': ' can', 'lang': 'es'}, {'char': 'ciones', 'lang': 'es'}, {'char': ' porque', 'lang': 'es'}, {'char': ' es', 'lang': 'es'}, {'char': ' como', 'lang': 'es'}]

Langs words  [{'word': ' o', 'lang': 'es'}, {'word': ' que', 'lang': 'es'}, {'word': ' es', 'lang': 'es'}, {'word': ' demasiado', 'lang': 'es'}, {'word': ' ruidosa', 'lang': 'es'}, {'word': ' o', 'lang': 'es'}, {'word': ' que', 'lang': 'es'}, {'word': ' tiene', 'lang': 'es'}, {'word': ' palabras', 'lang': 'es'}, {'word': ' que', 'lang': 'es'}, {'word': ' no', 'lang': 'es'}, {'word': ' deberían', 'lang': 'es'}, {'word': ' de', 'lang': 'es'}, {'word': ' ponerlas', 'lang': 'es'}, {'word': ' en', 'lang': 'es'}, {'word': ' canciones', 'lang': 'es'}, {'word': ' porque', 'lang': 'es'}, {'word': ' es', 'lang': 'es'}, {'word': ' como', 'lang': 'es'}]
```

```python
import pytorch_lightning as pl
from omegaconf import OmegaConf, open_dict

from nemo.collections.asr.models import ASRModel


def main():

    trainer = pl.Trainer(gpus=[0], precision=32, logger=False)

    # Char model
    # path = "Path to char model"

    # Multilingual model
    path = "/data3/models/enes_agg_277e.nemo"

    asr_model = ASRModel.restore_from(path, trainer=trainer, map_location='cpu')  

    asr_model.eval()
    asr_model.freeze()

    asr_model.preprocessor.featurizer.dither = 0.0
    asr_model.preprocessor.featurizer.pad_to = 0

    decoding_cfg = OmegaConf.create({})

    with open_dict(decoding_cfg):
        decoding_cfg.strategy = "greedy_batch"
        decoding_cfg.preserve_alignments = True
        decoding_cfg.compute_langs = True

    asr_model.change_decoding_strategy(decoding_cfg)

    # speech sample 
    WAV = 'es.wav'

    hyp, _ = asr_model.transcribe([WAV],
                   batch_size=1, return_hypotheses=True)

    print("Evaluating ---")
    print()
    print("Transcription: ", hyp[0].text)
    print()
    print("Langs: ", hyp[0].langs)
    print()
    print("Langs chars", hyp[0].langs_chars)
    print()
    print("Langs words ", hyp[0].langs_words)
    print()


if __name__ == '__main__':
    main()
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/metrics/rnnt_wer.py [hunk 1]
```diff
@@ -368,6 +369,11 @@ def decode_hypothesis(self, hypotheses_list: List[Hypothesis]) -> List[Union[Hyp
             # De-tokenize the integer tokens
             hypotheses_list[ind].text = hypothesis
 
+            if self.compute_langs:
+                hypotheses_list[ind].langs = self.decode_tokens_to_lang(prediction)
```
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer.py:373]: Doesn't this only work if the Tokenizer is an aggregate one ? Please check it in the constructor and raise an error.

File: nemo/collections/asr/metrics/rnnt_wer.py [hunk 2]
```diff
@@ -397,6 +403,44 @@ def decode_ids_to_tokens(self, tokens: List[int]) -> List[str]:
         """
         raise NotImplementedError()
 
+    @abstractmethod
```
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer.py:406]: I'd rather not make it abstract, instead by default raising an error that it is unimplemented. 

Subclass must check if tokenizer supports it, then return something, else error out

File: nemo/collections/asr/metrics/rnnt_wer.py [hunk 3]
```diff
@@ -397,6 +403,44 @@ def decode_ids_to_tokens(self, tokens: List[int]) -> List[str]:
...
+        Args:
+            tokens: List of int representing the token ids.
+
+        Returns:
+            A list of decoded LIDS.
+        """
+        raise NotImplementedError()
+
+    def decode_ids_to_words_and_langs(self, tokens: List[int]) -> List[str]:
+        """
```
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer.py:433]: This is gonna cause combinatorial explosion and is unnecessary imo. Try to remove the need for this flag by pre checks, and simply augmenting the output if necessary?

May not be possible so this is not necessary if it's not possible.

File: nemo/collections/asr/metrics/rnnt_wer.py [hunk 4]
```diff
@@ -812,6 +856,45 @@ def decode_ids_to_tokens(self, tokens: List[int]) -> List[str]:
...
+        Compute the most likely language ID (LID) string given the tokens.
+
+        Args:
+            tokens: List of int representing the token ids.
+
+        Returns:
+            A decoded LID string.
+        """
+        lang = self.tokenizer.ids_to_lang(tokens)
```
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer.py:869]: Check then return else error

File: nemo/collections/asr/metrics/rnnt_wer_bpe.py [hunk 5]
```diff
@@ -157,6 +157,46 @@ def decode_ids_to_tokens(self, tokens: List[int]) -> List[str]:
...
+        """
+        Compute the most likely language ID (LID) string given the tokens.
+
+        Args:
+            tokens: List of int representing the token ids.
+
+        Returns:
+            A decoded LID string.
+        """
+        lang = self.tokenizer.ids_to_lang(tokens)
```
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer_bpe.py:170]: Same for all these methods

File: nemo/collections/common/tokenizers/aggregate_tokenizer.py [hunk 6]
```diff
@@ -147,6 +151,85 @@ def ids_to_tokens(self, ids):
 
         return tokens
 
+    def ids_to_langs(self, ids):
```
Comment (titu1994) [inline nemo/collections/common/tokenizers/aggregate_tokenizer.py:154]: All of these methods are part of decoding - is it practical that user sees these methods ? Can we hide the internal ones or do we expect users to often use so many public methods

**Review by titu1994 — CHANGES_REQUESTED** [review]
Very cool PR !

File: nemo/collections/asr/metrics/rnnt_wer.py (hunk 2 above)
Comment (bmwshop) [reply nemo/collections/asr/metrics/rnnt_wer.py]: addressed

File: nemo/collections/asr/metrics/rnnt_wer.py (hunk 4 above)
Comment (bmwshop) [reply nemo/collections/asr/metrics/rnnt_wer.py]: addressed

File: nemo/collections/common/tokenizers/aggregate_tokenizer.py (hunk 6 above)
Comment (bmwshop) [reply nemo/collections/common/tokenizers/aggregate_tokenizer.py]: addressed

File: nemo/collections/asr/metrics/rnnt_wer_bpe.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/metrics/rnnt_wer_bpe.py:177]: ++

File: nemo/collections/common/tokenizers/aggregate_tokenizer.py [hunk 7]
```diff
@@ -76,18 +76,21 @@ def __init__(self, tokenizers: Dict):
...
-        # if there are two tokenizers, [0,1] each with 128 tokens, the aggregate tokenizer
-        # token range will be [0,255]. The below method provides two look up tables:
+        # if there are two tokenizers, [0,1], ['en', 'es'], each with 128 tokens, the aggregate tokenizer
+        # token range will be [0,255]. The below method provides three look up tables:
         # one, to convert the incoming token id -- e.g. 200 into its real id (200-127 = 73)
-        # and to compute the tokenizer id that should process that token (1)
-        offset_token_ids_by_token_id, tokenizers_by_token_id = self._calculate_offsets()
+        # second, to compute the tokenizer id that should process that token (1)
+        # third, the compute the lang id for that token ('es')
+        offset_token_ids_by_token_id, tokenizers_by_token_id, langs_by_token_id = self._calculate_offsets()
```
Comment (titu1994) [inline nemo/collections/common/tokenizers/aggregate_tokenizer.py:84]: Split into different lines

**Review by fayejf — CHANGES_REQUESTED** [review]
This is really cool. 

is it possible to merge your example code snippet to example scripts? say `transcribe_speech.py`? 

How about test?

File: examples/asr/transcribe_speech.py [hunk 8]
```diff
@@ -251,7 +277,9 @@ def autocast():
             with open(cfg.dataset_manifest, 'r') as fr:
                 for idx, line in enumerate(fr):
                     item = json.loads(line)
-                    item['pred_text'] = transcriptions[idx]
+                    item['pred_text'] = transcriptions[idx].text
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:280]: Isn't this only if hypothesis objects is returned ?

File: nemo/collections/asr/metrics/rnnt_wer_bpe.py [hunk 9]
```diff
@@ -157,6 +164,80 @@ def decode_ids_to_tokens(self, tokens: List[int]) -> List[str]:
...
+        return lang
+
+    def decode_ids_to_langs(self, tokens: List[int]) -> List[str]:
```
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer_bpe.py:180]: Too many public methods. Is it really necessary ?

File: nemo/collections/asr/metrics/rnnt_wer_bpe.py [hunk 10]
```diff
@@ -157,6 +164,80 @@ def decode_ids_to_tokens(self, tokens: List[int]) -> List[str]:
...
+        Args:
+            tokens: List of int representing the token ids.
+
+        Returns:
+            A list of decoded LIDS.
+        """
+        lang_list = self.tokenizer.ids_to_text_and_langs(tokens)
+        return lang_list
+
+    def decode_ids_to_words_and_langs(self, tokens: List[int]) -> List[str]:
```
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer_bpe.py:193]: Why is this necessary ? We won't have one method for every single combination like this. Try to decompose operations and then build step by step

**Review by titu1994 — CHANGES_REQUESTED** [review]

File: examples/asr/transcribe_speech.py (hunk 8 above)
Comment (bmwshop) [reply examples/asr/transcribe_speech.py]: sorry. fixed.

File: nemo/collections/asr/metrics/rnnt_wer.py (hunk 1 above)
Comment (bmwshop) [reply nemo/collections/asr/metrics/rnnt_wer.py]: fixed

File: nemo/collections/asr/metrics/rnnt_wer_bpe.py [hunk 11]
```diff
@@ -125,6 +131,7 @@ class RNNTBPEDecoding(AbstractRNNTDecoding):
     def __init__(self, decoding_cfg, decoder, joint, tokenizer: TokenizerSpec):
         blank_id = tokenizer.tokenizer.vocab_size
         self.tokenizer = tokenizer
+        self.compute_langs = decoding_cfg.get('compute_langs', False)
```
[the remark this answers is not in the document]
Comment (bmwshop) [reply nemo/collections/asr/metrics/rnnt_wer_bpe.py:134]: fixed

File: nemo/collections/asr/metrics/rnnt_wer_bpe.py (hunk 10 above)
Comment (bmwshop) [reply nemo/collections/asr/metrics/rnnt_wer_bpe.py]: fixed

File: examples/asr/transcribe_speech.py [hunk 12]
```diff
@@ -149,12 +156,24 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
...
 
+    # we will adjust this flag is the model does not support it
+    compute_langs = cfg.compute_langs
+
     # Setup decoding strategy
     if hasattr(asr_model, 'change_decoding_strategy'):
         # Check if ctc or rnnt model
         if hasattr(asr_model, 'joint'):  # RNNT model
-            asr_model.change_decoding_strategy(cfg.rnnt_decoding)
+            if isinstance(asr_model.decoding.tokenizer, AggregateTokenizer):
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:166]: I was just considering, would it be bad if we made return of the hypothesis the default thing and ignore the return of pure text case ? I want to keep this script as simple as possible, and it's getting messy with all these ifs and else's. 

The user can set the flag and run inference, let the code hard crash if it's not supported in the lower level call. Skip this if else and logging warning + explicitly updating the config etc.

File: examples/asr/transcribe_speech.py [hunk 13]
```diff
@@ -212,6 +231,11 @@ def autocast():
         return cfg
 
     # transcribe audio
+
+    return_hypotheses = False
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:235]: Make return hypothesis set to true by default. High time we moved to the more general interface. Can you update the rest of the script to use Hypothesis outputs ? It's not technically part of LID but might as well do the cleanup in this pr to group related changes.

File: examples/asr/transcribe_speech.py [hunk 14]
```diff
@@ -241,17 +271,34 @@ def autocast():
...
         transcriptions = transcriptions[0]
+
     # write audio transcriptions
     with open(cfg.output_filename, 'w', encoding='utf-8') as f:
         if cfg.audio_dir is not None:
-            for idx, text in enumerate(transcriptions):
+            for idx, transcription in enumerate(transcriptions):
+                text = transcription
+                if return_hypotheses:
+                    text = transcription.text
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:281]: This will be new default

File: examples/asr/transcribe_speech.py [hunk 15]
```diff
@@ -241,17 +271,34 @@ def autocast():
...
+                    item['pred_lang_chars'] = transcription.langs_chars
+
                 f.write(json.dumps(item) + "\n")
         else:
             with open(cfg.dataset_manifest, 'r') as fr:
                 for idx, line in enumerate(fr):
                     item = json.loads(line)
-                    item['pred_text'] = transcriptions[idx]
+                    if return_hypotheses:
+                        item['pred_text'] = transcriptions[idx].text
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:294]: No need for if else's, Hypothesis returned can be new default

File: nemo/collections/asr/metrics/rnnt_wer.py [hunk 16]
```diff
@@ -936,6 +990,9 @@ class RNNTDecodingConfig:
     # compute RNNT time stamps
     compute_timestamps: Optional[bool] = None
 
+    # compute language IDs
+    compute_langs: Optional[bool] = None
```
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer.py:994]: If it's bool, why optional ?

Comment (titu1994) [reply nemo/collections/asr/metrics/rnnt_wer.py:994]: I mean just default it to false ?

File: nemo/collections/asr/metrics/rnnt_wer_bpe.py (hunk 11 above)
Comment (titu1994) [inline nemo/collections/asr/metrics/rnnt_wer_bpe.py:134]: You can move this line to AbstractRNNTDecoding

**Review by titu1994 — COMMENTED** [review]
Overall looks much better. 2 request : 
- let's remove if else checks and let the script crash if incorrect args are provided. 
- Also make hypotheses object returned by default.

File: nemo/collections/asr/metrics/rnnt_wer.py (hunk 16 above)
Comment (bmwshop) [reply nemo/collections/asr/metrics/rnnt_wer.py]: You're right and done

File: nemo/collections/asr/metrics/rnnt_wer_bpe.py (hunk 11 above)
Comment (bmwshop) [reply nemo/collections/asr/metrics/rnnt_wer_bpe.py]: done

File: examples/asr/transcribe_speech.py (hunk 12 above)
Comment (bmwshop) [reply examples/asr/transcribe_speech.py]: done

File: examples/asr/transcribe_speech.py (hunk 13 above)
Comment (bmwshop) [reply examples/asr/transcribe_speech.py]: done

File: examples/asr/transcribe_speech.py [hunk 17]
```diff
@@ -149,12 +156,29 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
...
         if hasattr(asr_model, 'joint'):  # RNNT model
-            asr_model.change_decoding_strategy(cfg.rnnt_decoding)
+            if hasattr(asr_model.decoding, 'tokenizer') and isinstance(
+                asr_model.decoding.tokenizer, AggregateTokenizer
+            ):
+                rnnt_decoding = RNNTDecodingConfig(fused_batch_size=-1, compute_langs=cfg.compute_langs)
+                asr_model.change_decoding_strategy(rnnt_decoding)
+            else:
+                logging.warning("compute_langs flag turned on but model tokenizer is not Aggregate. Ignoring")
+                compute_langs = False
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:176]: Remove this else block entirely.

File: examples/asr/transcribe_speech.py [hunk 18]
```diff
@@ -149,12 +156,29 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
...
+                asr_model.change_decoding_strategy(cfg.rnnt_decoding)
         else:
+            if cfg.compute_langs:
+                logging.warning("CTC models do not support compute_langs flag for now. Ignoring.")
+                compute_langs = False
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:181]: Remove this too.

File: examples/asr/transcribe_speech.py [hunk 19]
```diff
@@ -149,12 +156,29 @@ def main(cfg: TranscriptionConfig) -> TranscriptionConfig:
...
+    # we will adjust this flag is the model does not support it
+    compute_langs = cfg.compute_langs
+
     # Setup decoding strategy
     if hasattr(asr_model, 'change_decoding_strategy'):
         # Check if ctc or rnnt model
         if hasattr(asr_model, 'joint'):  # RNNT model
-            asr_model.change_decoding_strategy(cfg.rnnt_decoding)
+            if hasattr(asr_model.decoding, 'tokenizer') and isinstance(
+                asr_model.decoding.tokenizer, AggregateTokenizer
```
Comment (titu1994) [inline examples/asr/transcribe_speech.py:170]: Revert thd entire if else block.

**Review by titu1994 — CHANGES_REQUESTED** [review]

**Review by titu1994 — APPROVED** [review]
Looks great ! Thanks !

**Review by fayejf — APPROVED** [review]
LGTM! Thank you!
